viewer/packages/tools/src/Timeline/TimelineTool.ts

```diff
@@ -0,0 +1,203 @@
+/*!
+ * Copyright 2021 Cognite AS
+ */
+
+import TWEEN from '@tweenjs/tween.js';
+
+import { Cognite3DModel } from '@reveal/core';
+import { Cognite3DViewerToolBase } from '../Cognite3DViewerToolBase';
+import { Keyframe } from './Keyframe';
+import { TimelineDateUpdateDelegate } from './types';
+import { EventTrigger, assertNever } from '@reveal/core/utilities';
+
+/**
+ * Tool to applying styles to nodes based on date to play them over in Timeline
+ */
+export class TimelineTool extends Cognite3DViewerToolBase {
+  private readonly _model: Cognite3DModel;
+  private _keyframes: Keyframe[];
+  private _playback: TWEEN.Tween | undefined = undefined;
+  private readonly _events = { dateChanged: new EventTrigger<TimelineDateUpdateDelegate>() };
+
+  constructor(cadModel: Cognite3DModel) {
+    super();
+
+    this._model = cadModel;
+    this._keyframes = new Array<Keyframe>();
+  }
+
+  /**
+   * Subscribe to the Date changed event
+   * @param event `dateChanged` event
+   * @param listener Listen to Timeline date Update during Playback
+   */
+  public subscribe(event: 'dateChanged', listener: TimelineDateUpdateDelegate): void {
+    switch (event) {
+      case 'dateChanged':
+        this._events.dateChanged.subscribe(listener);
+        break;
+      default:
+        assertNever(event, `Unsupported event: '${event}'`);
+    }
+  }
+
+  /**
+   * Unsubscribe to the Date changed event
+   * @param event `dateChanged` event
+   * @param listener Remove Listen to Timeline date Update
+   */
+  public unsubscribe(event: 'dateChanged', listener: TimelineDateUpdateDelegate): void {
+    switch (event) {
+      case 'dateChanged':
+        this._events.dateChanged.unsubscribe(listener as TimelineDateUpdateDelegate);
+        break;
+      default:
+        assertNever(event, `Unsupported event: '${event}'`);
+    }
+  }
+
+  /**
+   * Create Key frame for the Timeline
+   * @param date - date value by Date.now() since January 1, 1970
+   */
+  public createKeyframe(date: Date): Keyframe {
+    const keyframe = new Keyframe(this._model, date);
+    this._keyframes.push(keyframe);
+    this.sortKeyframesByDates();
+
+    return keyframe;
+  }
+
+  /**
+   * Returns the keyframe at the date given, or undefined if not found.
+   * @param date
+   * @returns
+   */
+  public getKeyframeByDate(date: Date): Keyframe | undefined {
+    return this._keyframes.find(candidate => candidate.getKeyframeDate() === date);
+  }
+
+  /**
+   * Removes the Keyframe from the timeline. Does nothing if the keyframe isn't part of the timeline.
+   * @param keyframe - Keyframe to be removed from the timeline
+   */
+  public removeKeyframe(keyframe: Keyframe) {
+    if (this._keyframes.length > 0) {
+      const index = this._keyframes.findIndex(obj => obj === keyframe);
+
+      if (index > -1) {
+        this._keyframes = this._keyframes.splice(index, 1);
+      }
+    }
```

Comment:
```suggestion
    const index = this._keyframes.findIndex(obj => obj === keyframe);

    if (index > -1) {
      this._keyframes = this._keyframes.splice(index, 1);
  }
```

Reply:
There's no need to check for the case where there are no keyframes - it's handled by the checking if the element is found. You can remove this in other functions as well, please look over

(Because less code is better than more code)